### PR TITLE
Validate inputs to Zonemaster::LDNS::RR::NSEC3::covers()

### LINKED ISSUE #174: Manipulating malformed NSEC3 RRs causes Perl to crash
Calling the `Zonemaster::LDNS::RR::NSEC3::covers` method on `Zonemaster::LDNS::RR::NSEC3` objects instantiated from NSEC3 resource records whose “next hashed owner name” RDATA field is empty causes Perl to crash with the following output:

```
perl: ./rdata.c:33: ldns_rdf_get_type: Assertion `rd != NULL' failed.
Aborted
```

## Example

The following examples use a version of `Zonemaster::LDNS` linked against an external custom build of `ldns`, version 1.8.3, compiled with debugging symbols.

Script:
```perl
#!/usr/bin/env perl

use strict;
use v5.10;

use Zonemaster::LDNS;

# This will crash
say Zonemaster::LDNS::RR->new(
    q{one.example. 0 IN NSEC3 \# 15 01 00 0001 01 AB 00 0006 400000000002}
)->covers("example");
```

Backtrace:
```
Program received signal SIGABRT, Aborted.
__GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
50      ../sysdeps/unix/sysv/linux/raise.c: No such file or directory.
(gdb) bt
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
#1  0x00007ffff7c6c859 in __GI_abort () at abort.c:79
#2  0x00007ffff7c6c729 in __assert_fail_base (fmt=0x7ffff7e02588 "%s%s%s:%u: %s%sAssertion `%s' failed.\n%n",
    assertion=0x7ffff769e27a "rd != NULL", file=0x7ffff769e270 "./rdata.c", line=33, function=<optimized out>)
    at assert.c:92
#3  0x00007ffff7c7dfd6 in __GI___assert_fail (assertion=0x7ffff769e27a "rd != NULL", file=0x7ffff769e270 "./rdata.c",
    line=33, function=0x7ffff769e520 <__PRETTY_FUNCTION__.24937> "ldns_rdf_get_type") at assert.c:101
#4  0x00007ffff767edc5 in ldns_rdf_get_type (rd=0x0) at ./rdata.c:33
#5  0x00007ffff765409a in ldns_dname_cat (rd1=0x0, rd2=0x555555a849d0) at ./dname.c:96
#6  0x00007ffff76585e0 in ldns_nsec_covers_name (nsec=0x555555a84990, name=0x5555558d5ec0) at ./dnssec.c:1530
#7  0x00007ffff76c2b80 in XS_Zonemaster__LDNS__RR__NSEC3_covers (my_perl=<optimized out>, cv=<optimized out>)
    at src/LDNS.xs:2355
#8  0x000055555565da10 in Perl_pp_entersub ()
#9  0x0000555555654166 in Perl_runops_standard ()
#10 0x00005555555c791c in perl_run ()
#11 0x000055555559d432 in main ()
```

### PR DESCRIPTION
**Pull request by marc-vanderwal**
## Purpose

This PR adds some input validation to the Zonemaster::LDNS::RR::NSEC3 package’s `covers` method.

## Context

Fix #174 and #175.

## Changes

* Add unit tests involving bad NSEC3 resource records and bad uses of Zonemaster::LDNS::RR::NSEC3::covers().
* Ensure the `covers()` method returns `undef`, instead of crashing, in the following cases:
  * The NSEC3 resource record has an empty next hashed owner name field in its RDATA: a malformed NSEC3 record cannot prove the absence of anything,
  * The `covers()` method is called with the root zone as argument: the root zone always exists and no NSEC3 record can ever prove its absence.

## How to test this PR

Unit tests were added, so all unit tests should pass.

### CODE REVIEW

**Review by matsduf — COMMENTED** [review]
Does this change mean that Zonemaster will not see such NSEC3 records? If so, it cannot report it to the user.

File: t/dnssec.t [hunk 1]
```diff
@@ -59,10 +59,28 @@ my $nsec = Zonemaster::LDNS::RR->new('xx.se.			7200	IN	NSEC	xx0r.se. NS RRSIG NS
...
 my $nsec3 = Zonemaster::LDNS::RR->new('NR2E513KM693MBTNVHH56ENF54F886T0.com. 86400 IN NSEC3 1 1 0 - NR2FUHQVR56LH70L6F971J3L6N1RH2TU NS DS RRSIG');
 isa_ok($nsec3, 'Zonemaster::LDNS::RR::NSEC3');
 ok($nsec3->covers('xx-example.com'), 'Covers xx-example.com');
 
+ok(!$nsec3->covers('.'), 'Does not cover the root domain');
+
+subtest 'malformed NSEC3 do not cause crashes' => sub {
+    # malformed because of empty hashed next owner name
+    my $example = Zonemaster::LDNS::RR->new(
+        q{example. 0 IN NSEC3 \# 15 01 00 0001 01 AB 00 0006 400000000002}
```
[the remark this answers is not in the document]
Comment (marc-vanderwal) [reply t/dnssec.t:73]: True, but I was looking for a minimal crashing example. The root cause of the problem was the empty hashed next owner name field in the RDATA, not the owner name.

**Comment by marc-vanderwal** [comment]
> Does this change mean that Zonemaster will not see such NSEC3 records? If so, it cannot report it to the user.

The changes I made make the `covers` method return `undef` when the NSEC3 record is not well-formed or when asked if an NSEC3 record covers the root zone. A bad NSEC3 record cannot prove the non-existence of anything. My main concern was addressing the crashing of Zonemaster due to these bad NSEC3 records, but I can amend the unit tests a bit to ensure that `undef` is returned in those edge cases.

You can also try running a test on `vsan.re` (thanks to @pnax for discovering that problem zone), where the DNSSEC10 test crashes. Now it gives an error.

Here’s what happens normally:
```
$ zonemaster-cli --show-module --show-testcase --test dnssec vsan.re
Seconds Level    Module       Testcase       Message
======= ======== ============ ============== =======
   0.79 ERROR    DNSSEC       DNSSEC01       DS record for zone vsan.re with keytag 24195 was created by digest algorithm 1 (SHA-1) which is deprecated. Fetched from the nameservers with IP addresses "193.176.144.22;194.0.36.1;194.0.9.1;194.146.106.46;2001:678:4c::1;2001:678:c::1;2001:67c:1010:11::53;2a00:d78:0:102:193:176:144:22".
   0.79 NOTICE   DNSSEC       DNSSEC01       No DS record created by digest algorithm 2 (SHA-256) is present for zone vsan.re.
perl: ./rdata.c:33: ldns_rdf_get_type: Assertion `rd != NULL' failed.
Aborted
```

And after the fix:
```
$ zonemaster-cli --show-module --show-testcase --test dnssec vsan.re
Seconds Level     Module       Testcase       Message
======= ========= ============ ============== =======
   0.88 ERROR     DNSSEC       DNSSEC01       DS record for zone vsan.re with keytag 24195 was created by digest algorithm 1 (SHA-1) which is deprecated. Fetched from the nameservers with IP addresses "193.176.144.22;194.0.36.1;194.0.9.1;194.146.106.46".
   0.88 NOTICE    DNSSEC       DNSSEC01       No DS record created by digest algorithm 2 (SHA-256) is present for zone vsan.re.
   6.24 ERROR     DNSSEC       DNSSEC10       A non-existent name is not correctly covered by the NSEC3 records. Fetched from the nameservers with IP addresses "81.26.208.160;81.26.219.100".
```

File: t/dnssec.t (hunk 1 above)
[the remark this answers is not in the document]
Comment (matsduf) [reply t/dnssec.t:73]: Maybe a comment on that?

**Review by matsduf — APPROVED** [review]

**Review by matsduf — APPROVED** [review]

File: t/dnssec.t (hunk 1 above)
[the remark this answers is not in the document]
Comment (marc-vanderwal) [reply t/dnssec.t:75]: I’ve improved the comments for the resource records.

**Review by mattias-p — APPROVED** [review]
